### PR TITLE
Release tools

### PR DESCRIPTION
**Pull request by razvan**
## Requirements

```
cargo install --git https://github.com/sunng87/cargo-release cargo-release
```

To test this before it's merged, you need to merge it locally to `main` first. This is because releases are only possible from `main` but the required files are not there yet. Once this PR is merged it will just work, of course.

After that, run `./release/scripts/perform.sh`

## What it does:
* creates a new branch called `release-YYYYmmDDHHMMSS`
* Bumps the version of all crates to the next minor version.
* Commits with a "release" message and tags it with this version
* Bumps the versions of all crates to the next development (nightly) version.
* Commits with a "prepare next release" message
* Pushes everything
* **Optional**: if you have the GitHub cli tools installed, it creates a PR.

## What it doesn't do:
* Changelog or readme updates.
* Doesn't publish crates.

## Review Checklist
- [ ] Code contains useful comments
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)

### CODE REVIEW

**Comment by soenkeliebau** [comment]
I've left a few comments on the PR against the monitoring operator. Should have probably reviewed here and not over there, sorry :(

File: Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# This file is automatically @generated by Cargo.
```
Comment (siegfriedweber) [inline Cargo.lock:1]: Cargo.lock should not be added to libraries.

File: release.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+consolidate-commits = true
+no-dev-version = false
+dev-version-ext = "nightly"
+pre-release-hook = ["../release/scripts/release-pre.sh"]
+pre-release-commit-message = "release"
```
Comment (siegfriedweber) [inline release.toml:6]: Why not `"release {{version}}"` as suggested in the documentation of cargo release?

File: release.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+post-release-commit-message = "start next development iteration"
```
Comment (siegfriedweber) [inline release.toml]: I do not like the word "iteration" in this context. What about `"prepare {{next_version}}"`?

File: release.toml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+disable-push = true
+disable-tag = true
+tag-name = "{{version}}"
```
Comment (siegfriedweber) [inline release.toml:10]: Does this also work for workspaces or do we have to prepend the tag with the according crate name?

File: release/scripts/perform.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    exit 1
+  fi
+
+  git switch -c ${RELEASE_BRANCH} ${BASE_BRANCH}
+  git push -u origin ${RELEASE_BRANCH}
+}
+
+tag_release_commit() {
+  local TAG=$1
+  git tag $TAG HEAD^
```
Comment (siegfriedweber) [inline release/scripts/perform.sh]: An annotated tag (`-a`) with a proper message should be created.

**Comment by razvan** [comment]
> I've left a few comments on the PR against the monitoring operator. Should have probably reviewed here and not over there, sorry :(

I've implemented the feedback in both PRs but only responded to the comments in the monitoring-operator PR.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]
I am not sure if the release script should be replicated in all projects. It would be easier to update it if it is a tool located in an own repository.

Currently it does not work on my machine because cargo release complains that a file or directory does not exist (I tested it on the agent-integration-tests):

```
Branch 'release-20210820075317' set up to track remote branch 'release-20210820075317' from 'origin'.
[2021-08-20T07:53:19Z INFO ] Update agent-integration-tests to version 0.2.0
[2021-08-20T07:53:19Z WARN ] Fatal: IO Error: No such file or directory (os error 2)
```
I will try to find out the reason.

**Update:**

cargo release could not find the pre-release-hook. The following change works for me:

```diff
- pre-release-hook = ["../release/scripts/release-pre.sh"]
+ pre-release-hook = ["release/scripts/release-pre.sh"]
```

Answer by @razvan:
This will not work for the operators because the hooks are executed in the same folders as the crates.

File: release.toml (hunk 2 above)
Comment (razvan) [reply release.toml:6]: I tried it, it didn't interpolate the {{version}} string. Looks like a bug.

File: release.toml (hunk 3 above)
Comment (razvan) [reply release.toml]: See above.

File: release.toml (hunk 4 above)
Comment (razvan) [reply release.toml:10]: It works

File: release/scripts/perform.sh (hunk 5 above)
Comment (razvan) [reply release/scripts/perform.sh]: Done.

File: release/scripts/release-pre.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
```
Comment (siegfriedweber) [inline release/scripts/release-pre.sh]: For systems where `bash` is not located in `/bin`:

```suggestion
#!/usr/bin/env bash
```

Comment (razvan) [reply release/scripts/release-pre.sh]: Done.

File: release.toml (hunk 2 above)
Comment (razvan) [reply release.toml:6]: I created an issue for it: https://github.com/sunng87/cargo-release/issues/310

File: Cargo.lock (hunk 1 above)
Comment (razvan) [reply Cargo.lock:1]: It's not added by the shell scripts. If it's added by `cargo-release` I'm afraid there is not much we can do, except maybe to add it to `.gitignore`.

**Comment by soenkeliebau** [comment]
> I am not sure if the release script should be replicated in all projects. It would be easier to update it if it is a tool located in an own repository.

Would we need to adapt the script to pass in a working directory in that case or would you clone the tool as a submodule / copy it to the repos before using it?
But overall I am in favor of having this once instead of adding it to every repo.

**Comment by siegfriedweber** [comment]
> > I am not sure if the release script should be replicated in all projects. It would be easier to update it if it is a tool located in an own repository.
> 
> Would we need to adapt the script to pass in a working directory in that case or would you clone the tool as a submodule / copy it to the repos before using it?
> But overall I am in favor of having this once instead of adding it to every repo.

I would create a tool/script which can be added to the `PATH` and which works on the current directory.

**Comment by soenkeliebau** [comment]
> I would create a tool/script which can be added to the `PATH` and which works on the current directory.

Do we need to distinguish between workspace repos and non workspace repos in that script? The current cargo release command in the shell script has a `--workspace' parameter, which I cannot find in the docs and am unsure what that does. Do we need that or can the module figure out itself if it is a workspace?

**Comment by lfrancke** [comment]
As far as I read that parameter just iterates over all workspace members and calls cargo release for each crate.

**Comment by soenkeliebau** [comment]
So we would indeed need to check whether it is a workspace in our script and specify or not specify that parameter accordingly?

**Comment by razvan** [comment]
Closing because the release tools have been extracted to their own repository. See this draft PR : https://github.com/stackabletech/tools/pull/1

**Pull request closed without merging**